python/daqconf/core/fragment_producers.py

```diff
@@ -25,16 +26,77 @@ def set_mlt_links(the_system, mlt_app_name="trigger", verbose=False):
     lives in an application with name `mlt_app_name` and has the name
     "mlt".
     """
-    mlt_links = []
+
+    ### additional mapping to allow ROI readout
+    mlt_readout_map = {}
+    for trigger_sid,conf in tp_infos.items():
+        if isinstance(conf, TAInfo):
+            for key in mlt_readout_map.keys():
+                if mlt_readout_map[key]["group"] == conf.region_id:
+                    mlt_readout_map[key]["elements"].append(trigger_sid )
+        elif isinstance(conf, TPInfo):
+            for key in mlt_readout_map.keys():
+                if key == conf.tp_ru_sid:
+                    mlt_readout_map[key]["group"] = conf.region_id
+                    mlt_readout_map[key]["elements"].append(conf.tp_ru_sid)
+                    mlt_readout_map[key]["elements"].append(trigger_sid)
+        elif isinstance(conf, TCInfo):
+            mlt_map_entry = { "group": -1, "elements": [trigger_sid] }
+            mlt_readout_map[-1] = mlt_map_entry
+        else:
+            # readout unit
+            mlt_map_entry = { "group": trigger_sid, "elements": [] }
+            for stream in conf.streams:
+                mlt_map_entry["elements"].append( stream.src_id )
+            mlt_readout_map[trigger_sid] = mlt_map_entry
+
+    if verbose:
+        console.log(f"MLT Readout Map: {mlt_readout_map}")
+
+    mlt_links = {"mandatory": [], "groups": []}
+    for key in mlt_readout_map.keys():
+        if key != -1:
+            mlt_links["groups"].append({"group": key, "links": []})
+    
     for producer in the_system.get_fragment_producers():
         if producer.is_mlt_producer:
             source_id = producer.source_id
-            mlt_links.append( mlt.SourceID(subsystem=ensure_subsystem_string(source_id.subsystem), element=source_id.id) )
+            # case for HSI
+            if source_id.subsystem == SourceID.Subsystem.kHwSignalsInterface:
+                mlt_links["mandatory"].append( mlt.SourceID(subsystem=ensure_subsystem_string(source_id.subsystem), element=source_id.id) )
+            # case for RU
+            elif source_id.subsystem == SourceID.Subsystem.kDetectorReadout:
+                matched = False
+                for key in mlt_readout_map.keys():
+                    if key != -1:
+                        if source_id.id in mlt_readout_map[key]["elements"]:
+                            mlt_links["groups"][key]["links"].append( mlt.SourceID(subsystem=ensure_subsystem_string(source_id.subsystem), element=source_id.id) )
+                            matched = True
+                # special case to cover readout that is not a TP source for trigger
+                if not matched:
+                    mlt_links["groups"].append({"group": len(mlt_links["groups"]), "links": []})
+                    mlt_links["groups"][len(mlt_links["groups"])-1]["links"].append( mlt.SourceID(subsystem=ensure_subsystem_string(source_id.subsystem), element=source_id.id) )
```

Comment:
No "mandatory" case here? Is there documentation about this that I could read?

Reply:
All currently existing documentation is here: [trigger Issue 230](https://github.com/DUNE-DAQ/trigger/issues/230)